arraycontext/pytest.py

```diff
@@ -28,48 +29,124 @@
 THE SOFTWARE.
 """
 
+from typing import Optional, Sequence
 
-# {{{ pytest integration
+import pyopencl as cl
+from pyopencl.tools import _ContextFactory
 
-def pytest_generate_tests_for_pyopencl_array_context(metafunc):
-    """Parametrize tests for pytest to use a
-    :class:`~arraycontext.PyOpenCLArrayContext`.
 
-    Performs device enumeration analogously to
-    :func:`pyopencl.tools.pytest_generate_tests_for_pyopencl`.
+# {{{ array context factories
 
-    Using the line:
+class _PyOpenCLArrayContextFactory(_ContextFactory):
+    force_device_scalars = True
+
+    def __call__(self):
+        ctx = super().__call__()
+        from arraycontext.impl.pyopencl import PyOpenCLArrayContext
+        return PyOpenCLArrayContext(
+                cl.CommandQueue(ctx),
+                force_device_scalars=self.force_device_scalars)
+
+    def __str__(self):
+        return ("<array context factory for <pyopencl.Device '%s' on '%s'>" %
+                (self.device.name.strip(),
+                 self.device.platform.name.strip()))
+
+
+class _DeprecatedPyOpenCLArrayContextFactory(_PyOpenCLArrayContextFactory):
+    force_device_scalars = False
+
+
+_ARRAY_CONTEXT_FACTORY_DICT = {
+        "pyopencl": _PyOpenCLArrayContextFactory,
+        }
+_ALL_ARRAY_CONTEXT_FACTORY_DICT = {
+        "pyopencl-deprecated": _DeprecatedPyOpenCLArrayContextFactory,
+        }
+
+_ALL_ARRAY_CONTEXT_FACTORY_DICT.update(_ARRAY_CONTEXT_FACTORY_DICT)
+
+# }}}
+
+
+# {{{ pytest integration
+
+def pytest_generate_tests_for_array_context(
+        metafunc,
+        impls: Optional[Sequence[str]] = None):
```

Comment:
I like the strings for their brevity, but if things go according to plan, we'll be awash in array contexts before too long, all of which will need testing. (For example, meshmode will have its own subclasses of array contexts before too long, cf. https://github.com/inducer/arraycontext/pull/29.) So I think documenting the interface to the factories and making it easy to make new pyopencl-using factories is probably a good move.

Reply:
What do you think about something like https://github.com/inducer/arraycontext/pull/44#discussion_r657370998?

Both approaches seem like they would do the job without much hassle, so I don't feel strongly either way.

Reply:
Added a `PytestArrayContextFactory` class in there with some documentation in a77105e. Also updated `pytest_generate_tests_for_array_contexts` to work with a mix of those and strings.